lib/agent.js

```diff
@@ -27,7 +32,85 @@ module.exports = class HTTPAgent {
   }
 
   keepSocketAlive (socket) {
-    return false
+    if (this._keepAlive === -1) return false
+
+    socket.setKeepAlive(true, this._keepAlive)
+    socket.unref()
+
+    return true
+  }
+
+  getName (opts) {
+    return `${opts.host}:${opts.port}`
+  }
+
+  addRequest (req, opts) {
+    opts = { ...opts, ...this._opts }
+
+    const name = this.getName(opts)
+
+    let socket
+
+    if (this._freeSockets.has(name)) {
+      const sockets = this._freeSockets.get(name)
+      socket = sockets.values().next().value
+      sockets.delete(socket)
+      if (sockets.size === 0) this._freeSockets.delete(name)
+
+      this.reuseSocket(socket, req)
+    } else {
+      socket = this.createConnection(opts)
+
+      socket.on('free', () => this._onfree(socket, name))
+      socket.on('close', () => this._onremove(socket, name))
+      socket.on('end', () => socket.destroy())
```

Comment:
This shouldn't be necessary so something else is most likely off if it _is_ currently necessary.

Reply:
Removed the `socket.on('end', () => socket.destroy())` at https://github.com/holepunchto/bare-http1/pull/15/commits/7b27037279ad23baf14e3075a72356c1715caa42, but curiously, with that, several tests increased its duration by exactly 5 seconds.